### PR TITLE
[DRAFT] Refcount stats across scopes

### PR DESCRIPTION
**Pull request by ambuc**
Draft solution to https://github.com/envoyproxy/envoy/issues/2453

### CODE REVIEW

File: source/common/stats/BUILD [hunk 1]
```diff
@@ -19,6 +19,7 @@ envoy_cc_library(
         "//source/common/common:assert_lib",
         "//source/common/common:hash_lib",
         "//source/common/common:perf_annotation_lib",
+        "//source/common/common:shared_memory_hash_set_lib",
```
Comment (mrice32) [inline source/common/stats/BUILD]: test comment

File: source/common/stats/stats_impl.cc [hunk 2]
```diff
@@ -149,13 +149,48 @@ bool TagExtractorImpl::extractTag(const std::string& stat_name, std::vector<Tag>
   return false;
 }
 
+RawStatData* BlockRawStatDataAllocator::alloc(const std::string& name) {
+  absl::string_view key = name;
```
Comment (mrice32) [inline source/common/stats/stats_impl.cc]: Why not just use `name` directly throughout?

File: source/common/stats/stats_impl.h [hunk 3]
```diff
@@ -369,13 +372,26 @@ class HistogramImpl : public Histogram, public MetricImpl {
...
+ */
+class BlockRawStatDataAllocator : public RawStatDataAllocator {
+public:
+  BlockRawStatDataAllocator(const SharedMemoryHashSetOptions& options, bool init, uint8_t* memory) {
+    stats_set_ = std::make_unique<RawStatDataSet>(options, init, memory);
+  }
+
+  RawStatData* alloc(const std::string& name);
+  void free(RawStatData& data);
+  std::unique_ptr<RawStatDataSet> stats_set_;
```
Comment (mrice32) [inline source/common/stats/stats_impl.h]: Is there a reason we can't make this variable private?

File: source/server/hot_restart_impl.cc [hunk 4]
```diff
@@ -84,8 +85,8 @@ SharedMemory& SharedMemory::initialize(uint32_t stats_set_size, Options& options
   }
 
   // Stats::RawStatData must be naturally aligned for atomics to work properly.
-  RELEASE_ASSERT((reinterpret_cast<uintptr_t>(shmem->stats_set_data_) % alignof(RawStatDataSet)) ==
-                 0);
+  RELEASE_ASSERT(
```
Comment (mrice32) [inline source/server/hot_restart_impl.cc]: What changed here?

File: source/server/hot_restart_impl.cc [hunk 5]
```diff
@@ -115,15 +116,15 @@ std::string SharedMemory::version(size_t max_num_stats, size_t max_stat_name_len
 
 HotRestartImpl::HotRestartImpl(Options& options)
     : options_(options), stats_set_options_(sharedMemHashOptions(options.maxStats())),
-      shmem_(SharedMemory::initialize(RawStatDataSet::numBytes(stats_set_options_), options)),
+      shmem_(
+          SharedMemory::initialize(Stats::RawStatDataSet::numBytes(stats_set_options_), options)),
+      allocator_(stats_set_options_, options.restartEpoch() == 0, shmem_.stats_set_data_),
```
Comment (mrice32) [inline source/server/hot_restart_impl.cc]: You will probably need to initialize this below inside the scope where `lock` is created, which is where the stats set used to be initialized.  `allocator_` will probably need to be a unique_ptr as a result.  This is because the lock protects the initialization.  If you move it above to the initializer list, the lock won't be held while it's being constructed.  Side note: are you familiar with the idea of a scope lock (`std::unique_lock<Thread::BasicLockable>`)?

File: source/server/hot_restart_nop_impl.h [hunk 6]
```diff
@@ -14,7 +17,14 @@ namespace Server {
...
 public:
-  HotRestartNopImpl() {}
+  HotRestartNopImpl() {
+    int num_stats = 100;
+    const SharedMemoryHashSetOptions options = sharedMemHashOptions(num_stats);
+    ptr_ = static_cast<uint8_t*>(
+        calloc(SharedMemoryHashSet<Stats::RawStatData>::numBytes(options), num_stats));
+    stats_allocator_ = std::make_unique<Stats::BlockRawStatDataAllocator>(options, true, ptr_);
+  }
+  ~HotRestartNopImpl() { ::free(ptr_); }
```
Comment (mrice32) [inline source/server/hot_restart_nop_impl.h]: Be careful, here.  I'd be sure that that `stats_allocator_` doesn't attempt to write to this memory as it's destructing.  This is an artifact of the C++ order of destruction where the destructor runs first, and then the member variables are destroyed.  If this is a problem, we can modify the destructor to look like this:
```
~HotRestartNopImpl() { 
stats_allocator_ = nullptr;
::free(ptr_); 
}

File: test/common/stats/stats_impl_test.cc [hunk 7]
```diff
@@ -466,5 +468,35 @@ TEST(TagProducerTest, CheckConstructor) {
       "No regex specified for tag specifier and no default regex for name: 'test_extractor'");
 }
 
+TEST(SomeTest, HotAllocConsistencyTest) {
+  int max_stats_ = 16384; // ENVOY_DEFAULT_MAX_STATS
```
Comment (mrice32) [inline test/common/stats/stats_impl_test.cc]: If you want this to match that number, you should probably use the macro directly.

File: test/common/stats/stats_impl_test.cc [hunk 8]
```diff
@@ -466,5 +468,35 @@ TEST(TagProducerTest, CheckConstructor) {
...
+
+  SharedMemoryHashSetOptions options_ = Server::sharedMemHashOptions(max_stats_);
+  uint8_t* ptr = static_cast<uint8_t*>(
+      calloc(SharedMemoryHashSet<Stats::RawStatData>::numBytes(options_), max_stats_));
+  std::unique_ptr<RawStatDataSet> stats_set_ =
```
Comment (mrice32) [inline test/common/stats/stats_impl_test.cc]: Since you have your fancy new block allocator, why not use it here? :) .  Otherwise the test is really testing a reimplementation of your allocator rather than the allocator itself.

File: source/server/hot_restart_impl.cc (hunk 4 above)
Comment (ambuc) [reply source/server/hot_restart_impl.cc]: `RawStatDataSet` -> `Stats::RawStatDataSet`, which pushed it over some character limit and reflowed the formatting.

File: test/common/stats/stats_impl_test.cc (hunk 8 above)
Comment (ambuc) [reply test/common/stats/stats_impl_test.cc]: Good catch. Rewrote to look like
```
TEST(SomeTest, HotAllocConsistencyTest) { 
  int max_stats_ = 16384; // ENVOY_DEFAULT_MAX_STATS

  SharedMemoryHashSetOptions options_ = Server::sharedMemHashOptions(max_stats_);
  uint8_t* ptr = static_cast<uint8_t*>(
      calloc(SharedMemoryHashSet<Stats::RawStatData>::numBytes(options_), max_stats_));

  Stats::BlockRawStatDataAllocator allocator = 
    Stats::BlockRawStatDataAllocator(options_, true, ptr);                                                                                                                                                                      

  auto data_1 = allocator.alloc("foo");
  auto data_2 = allocator.alloc("foo");

  EXPECT_EQ(data_1, data_2);

  ::free(ptr);
}
```

File: test/common/stats/stats_impl_test.cc (hunk 7 above)
Comment (ambuc) [reply test/common/stats/stats_impl_test.cc]: I'll include `server/options_impl.h` and get access to it.

File: source/server/hot_restart_impl.cc (hunk 4 above)
Comment (mrice32) [reply source/server/hot_restart_impl.cc]: Ah, feel free to disregard.

File: source/common/stats/stats_impl.cc (hunk 2 above)
Comment (ambuc) [reply source/common/stats/stats_impl.cc]: `name` comes in as a `std::string&` and we want to be able to call things like `.remove_suffix()` on it, which are `absl::string_view` features.

File: source/server/hot_restart_impl.cc (hunk 5 above)
Comment (ambuc) [reply source/server/hot_restart_impl.cc]: This is a good catch. I reworked `allocator_` to be a unique pointer and constructed it not-inline, and changed all references to `allocator_.foo` to `allocator_->foo`.

File: source/server/hot_restart_nop_impl.h (hunk 6 above)
Comment (ambuc) [reply source/server/hot_restart_nop_impl.h]: I'll do this -- is this the kind of bug I can write a test for? Or is this just intuition that this could fail?

File: test/common/stats/stats_impl_test.cc [hunk 9]
```diff
@@ -466,5 +469,23 @@ TEST(TagProducerTest, CheckConstructor) {
       "No regex specified for tag specifier and no default regex for name: 'test_extractor'");
 }
 
+TEST(SomeTest, HotAllocConsistencyTest) {
```
Comment (ambuc) [inline test/common/stats/stats_impl_test.cc]: @mrice32 I think I don't have a great way to test this twice, once in the context of injecting shared memory, and once in the context of injecting non-shared memory. Is there a test we can write to prove this works in the original worry case in https://github.com/envoyproxy/envoy/issues/2453 ?

File: source/common/stats/stats_impl.cc (hunk 2 above)
Comment (mrice32) [reply source/common/stats/stats_impl.cc]: ahhh, makes sense.  I didn't realize you were substringing it.

File: source/server/hot_restart_nop_impl.h [hunk 10]
```diff
@@ -14,7 +17,18 @@ namespace Server {
...
 class HotRestartNopImpl : public Server::HotRestart {
 public:
-  HotRestartNopImpl() {}
+  HotRestartNopImpl() {
+    int num_stats = 100;
+    const MemoryHashSetOptions options = memHashOptions(num_stats);
+    ptr_ = static_cast<uint8_t*>(
+        calloc(MemoryHashSet<Stats::RawStatData>::numBytes(options), num_stats));
+    stats_allocator_ =
+        std::make_unique<Stats::BlockRawStatDataAllocator>(options, true, ptr_, &stat_lock_);
```
Comment (ambuc) [inline source/server/hot_restart_nop_impl.h]: @mrice32 when you see new statLock / stat_lock_ objects in HotRestart, HotRestartImpl, HotRestartNopImpl, and HotRestart mock objects, it's because of this line. I wanted to create and use a lock mechanism even when the restart isn't hot.

File: source/common/common/BUILD [hunk 11]
```diff
@@ -179,8 +179,8 @@ envoy_cc_library(
 )
 
 envoy_cc_library(
-    name = "shared_memory_hash_set_lib",
```
Comment (ambuc) [inline source/common/common/BUILD]: @mrice32 This is one of a few commits which rename `SharedMemoryHashSet` to `MemoryHashSet`, or `shared_memory_hash_set` to `memory_hash_set`, or `sharedMemOptions` to `memOptions`. I wanted to indicate that the MemoryHashSet doesn't need to be run on shared memory.

File: source/common/stats/stats_impl.cc [hunk 12]
```diff
@@ -149,13 +149,55 @@ bool TagExtractorImpl::extractTag(const std::string& stat_name, std::vector<Tag>
   return false;
 }
 
+RawStatData* BlockRawStatDataAllocator::alloc(const std::string& name) {
```
Comment (ambuc) [inline source/common/stats/stats_impl.cc]: I moved a lot of these tests from `hot_restart_impl_test.cc` since they are really unit tests on the allocator itself. A few stayed in `hot_restart_impl_test.cc` as they tested behavior across hot restart.

File: test/server/hot_restart_impl_test.cc [hunk 13]
```diff
@@ -111,120 +111,5 @@ TEST_F(HotRestartImplTest, crossAlloc) {
   EXPECT_EQ(stat5, stat5_prime);
 }
 
-TEST_F(HotRestartImplTest, truncateKey) {
```
Comment (ambuc) [inline test/server/hot_restart_impl_test.cc]: I moved a lot of these to `stats_impl.cc`. See the other comment here: https://github.com/ambuc/envoy/pull/1/files#r182548087

File: source/server/hot_restart_impl.cc [hunk 14]
```diff
@@ -140,36 +139,10 @@ HotRestartImpl::HotRestartImpl(Options& options)
 
 Stats::RawStatData* HotRestartImpl::alloc(const std::string& name) {
   // Try to find the existing slot in shared memory, otherwise allocate a new one.
-  std::unique_lock<Thread::BasicLockable> lock(stat_lock_);
```
Comment (ambuc) [inline source/server/hot_restart_impl.cc]: The whole functionality of `HotRestartImpl::alloc`, including the lock, got moved up into `BlockRawStatDataAllocator::alloc`. Same for `::free()`.

File: source/server/hot_restart_impl.cc [hunk 15]
```diff
@@ -472,23 +445,25 @@ void HotRestartImpl::terminateParent() {
 void HotRestartImpl::shutdown() { socket_event_.reset(); }
 
 std::string HotRestartImpl::version() {
-  return versionHelper(shmem_.maxStats(), Stats::RawStatData::maxNameLength(), *stats_set_);
```
Comment (ambuc) [inline source/server/hot_restart_impl.cc]: @mrice32 I wanted BlockRawStatDataAllocator's `stats_set_` member variable to be private, so I ended up writing a function here [1] to just fetch the version as a string.

[1] https://github.com/ambuc/envoy/pull/1/files#diff-fb83459430ccfd2f0d934dcfacdbd9d1R399

File: source/server/hot_restart_impl.cc [hunk 16]
```diff
@@ -472,23 +445,25 @@ void HotRestartImpl::terminateParent() {
...
-  const SharedMemoryHashSetOptions options = sharedMemHashOptions(max_num_stats);
-  const size_t bytes = RawStatDataSet::numBytes(options);
+  const MemoryHashSetOptions options = memHashOptions(max_num_stats);
+  const size_t bytes = Stats::RawStatDataSet::numBytes(options);
   std::unique_ptr<uint8_t[]> mem_buffer_for_dry_run_(new uint8_t[bytes]);
-  RawStatDataSet stats_set(options, true /* init */, mem_buffer_for_dry_run_.get());
+  Stats::RawStatDataSet stats_set(options, true /* init */, mem_buffer_for_dry_run_.get());
+  auto stats_set_version = stats_set.version();
 
-  return versionHelper(max_num_stats, max_stat_name_len, stats_set);
```
Comment (ambuc) [inline source/server/hot_restart_impl.cc]: See the comment here https://github.com/ambuc/envoy/pull/1/files#r182549180, we can now pass the string version instead of the whole set by reference.

File: source/server/hot_restart_impl.h [hunk 17]
```diff
@@ -204,12 +205,12 @@ class HotRestartImpl : public HotRestart,
   RpcBase* receiveRpc(bool block);
   void sendMessage(sockaddr_un& address, RpcBase& rpc);
   static std::string versionHelper(uint64_t max_num_stats, uint64_t max_stat_name_len,
-                                   RawStatDataSet& stats_set);
```
Comment (ambuc) [inline source/server/hot_restart_impl.h]: See https://github.com/ambuc/envoy/pull/1/files#r182549180

File: source/server/hot_restart_nop_impl.h [hunk 18]
```diff
@@ -14,7 +17,18 @@ namespace Server {
  */
 class HotRestartNopImpl : public Server::HotRestart {
 public:
-  HotRestartNopImpl() {}
+  HotRestartNopImpl() {
+    int num_stats = 100;
```
Comment (ambuc) [inline source/server/hot_restart_nop_impl.h]: @mrice32 TODO(jbuckland) find a less arbitrary number for this. I seem to recall setting this to `ENVOY_DEFAULT_MAX_STATS` caused testing to hang. What should we do here?

File: source/server/options_impl.cc [hunk 19]
```diff
@@ -13,11 +13,6 @@
 #include "spdlog/spdlog.h"
 #include "tclap/CmdLine.h"
 
-// Can be overridden at compile time
-#ifndef ENVOY_DEFAULT_MAX_STATS
```
Comment (ambuc) [inline source/server/options_impl.cc]: Moved to `options_impl.h` for visibility.

File: source/common/stats/stats_impl.h [hunk 20]
```diff
@@ -370,12 +373,44 @@ class HistogramImpl : public Histogram, public MetricImpl {
...
+   * optional stat_lock is passed in.
+   * @param options initialization parameters for MemoryHashSet
+   * @param init true if the shared-memory should be initialized on construction. See the
+   * MemoryHashSet constructor.
+   * @param memory the memory buffer for the set data.
+   * @param stat_lock an optional lock for thread safety. If nullptr, no locks are acquired.
+   * @return BlockRawStatDataAllocator newly constructed BlockRawStatDataAllocator.
+   */
+  BlockRawStatDataAllocator(const MemoryHashSetOptions& options, bool init, uint8_t* memory,
+                            Thread::BasicLockable* stat_lock)
```
Comment (mrice32) [inline source/common/stats/stats_impl.h]: Just a note: if we decide that we want this in all cases, we might consider making this a reference.  Or if we want this class to own it, we could pass in a unique_ptr and check that it's not nullptr.

File: source/server/hot_restart_impl.cc [hunk 21]
```diff
@@ -114,16 +115,14 @@ std::string SharedMemory::version(size_t max_num_stats, size_t max_stat_name_len
...
+          SharedMemory::initialize(Stats::RawStatDataSet::numBytes(stats_set_options_), options)),
       log_lock_(shmem_.log_lock_), access_log_lock_(shmem_.access_log_lock_),
       stat_lock_(shmem_.stat_lock_), init_lock_(shmem_.init_lock_) {
   {
-    // We must hold the stat lock when attaching to an existing shared-memory segment
-    // because it might be actively written to while we sanityCheck it.
-    std::unique_lock<Thread::BasicLockable> lock(stat_lock_);
-    stats_set_.reset(new RawStatDataSet(stats_set_options_, options.restartEpoch() == 0,
-                                        shmem_.stats_set_data_));
+    allocator_ = std::make_unique<Stats::BlockRawStatDataAllocator>(
```
Comment (mrice32) [inline source/server/hot_restart_impl.cc]: Why did you get rid of the locking here?

File: source/server/hot_restart_impl.cc [hunk 22]
```diff
@@ -140,36 +139,10 @@ HotRestartImpl::HotRestartImpl(Options& options)
...
-  std::unique_lock<Thread::BasicLockable> lock(stat_lock_);
-  ASSERT(data.ref_count_ > 0);
-  if (--data.ref_count_ > 0) {
-    return;
-  }
-  bool key_removed = stats_set_->remove(data.key());
-  ASSERT(key_removed);
-  memset(&data, 0, Stats::RawStatData::size());
-}
+void HotRestartImpl::free(Stats::RawStatData& data) { allocator_->free(data); }
```
Comment (mrice32) [inline source/server/hot_restart_impl.cc]: Why are we still forwarding these calls?

File: source/server/hot_restart_nop_impl.h (hunk 18 above)
Comment (mrice32) [reply source/server/hot_restart_nop_impl.h]: This is not a test, this is the real application, so this *must* be pulled from the supplied command line options/default just like the hot restart implementation does.  I would suggest supplying it via the constructor just like HotRestartImpl: 
```
HotRestartImpl(Options& options);
```

File: source/server/hot_restart_nop_impl.h (hunk 10 above)
Comment (mrice32) [reply source/server/hot_restart_nop_impl.h]: But why does it need to have its own accessor method?

File: test/common/stats/stats_impl_test.cc [hunk 23]
```diff
@@ -466,5 +472,148 @@ TEST(TagProducerTest, CheckConstructor) {
       "No regex specified for tag specifier and no default regex for name: 'test_extractor'");
 }
 
+class BlockRawStatDataAllocatorTest : public testing::Test {
+public:
+  void setup(int num_stats) {
+    Stats::RawStatData::configureForTestsOnly(options_);
+    MemoryHashSetOptions memory_hash_set_options_ = Server::memHashOptions(num_stats);
+    ptr_ = static_cast<uint8_t*>(
```
Comment (mrice32) [inline test/common/stats/stats_impl_test.cc]: We should probably move to the way Josh allocates the memory in his tests - see [here](https://github.com/envoyproxy/envoy/blob/master/test/common/common/shared_memory_hash_set_test.cc#L49).  Something like:

```
uint32_t num_bytes = MemoryHashSet<Stats::RawStatData>::numBytes(memory_hash_set_options_);
ptr_ = new uint8_t[num_bytes];
```

Side note: why are you multiplying the number of bytes times the number of stats?  That may be the source of your issues - you're squaring the capacity.

File: source/server/hot_restart_impl.cc (hunk 16 above)
Comment (mrice32) [reply source/server/hot_restart_impl.cc]: SGTM.

File: test/common/stats/stats_impl_test.cc (hunk 23 above)
Comment (ambuc) [reply test/common/stats/stats_impl_test.cc]: Great catch! I've updated this.

Comment (ambuc) [reply test/common/stats/stats_impl_test.cc]: Just kidding, I updated this for HotRestartNopImpl. I'll update it here too.

File: source/server/hot_restart_nop_impl.h [hunk 24]
```diff
@@ -5,6 +5,9 @@
 #include "envoy/server/hot_restart.h"
 
 #include "common/common/thread.h"
+#include "common/stats/stats_impl.h"
+
+#include "server/hot_restart_impl.h"
```
Comment (mrice32) [inline source/server/hot_restart_nop_impl.h]: Why does this include `hot_restart_impl.h`?

File: source/server/options_impl.cc (hunk 19 above)
Comment (mrice32) [reply source/server/options_impl.cc]: Makes sense.

File: test/server/hot_restart_nop_impl_test.cc [hunk 25]
```diff
@@ -0,0 +1,67 @@
...
+  Stats::RawStatData* stat2 = hot_restart_nop_->statsAllocator().alloc("stat2");
+  Stats::RawStatData* stat3 = hot_restart_nop_->statsAllocator().alloc("stat3");
+  Stats::RawStatData* stat4 = hot_restart_nop_->statsAllocator().alloc("stat4");
+  Stats::RawStatData* stat5 = hot_restart_nop_->statsAllocator().alloc("stat5");
+  hot_restart_nop_->statsAllocator().free(*stat2);
+  hot_restart_nop_->statsAllocator().free(*stat4);
+  stat2 = nullptr;
+  stat4 = nullptr;
+
+  EXPECT_CALL(options_, restartEpoch()).WillRepeatedly(Return(1));
```
Comment (mrice32) [inline test/server/hot_restart_nop_impl_test.cc]: Why is this here and why would the nopimpl call this method?

File: source/common/stats/stats_impl.cc (hunk 12 above)
Comment (mrice32) [reply source/common/stats/stats_impl.cc]: What tests are you referring to? This is an implementation file...?

File: source/server/hot_restart_impl.cc (hunk 15 above)
Comment (mrice32) [reply source/server/hot_restart_impl.cc]: Seems reasonable to me.

File: source/server/hot_restart_nop_impl.h (hunk 24 above)
Comment (ambuc) [reply source/server/hot_restart_nop_impl.h]: it needs `blockMemHashOptions` from there.

File: test/server/hot_restart_nop_impl_test.cc (hunk 25 above)
Comment (ambuc) [reply test/server/hot_restart_nop_impl_test.cc]: You're right, this is leftover from the `sameAlloc` function when it was being called on a more tightly coupled instance of HotRestart.

File: test/common/stats/thread_local_store_test.cc [hunk 26]
```diff
@@ -24,55 +28,24 @@ using testing::_;
 namespace Envoy {
 namespace Stats {
 
-/**
- * This is a heap test allocator that works similar to how the shared memory allocator works in
- * terms of reference counting, etc.
- */
-class TestAllocator : public RawStatDataAllocator {
```
Comment (ambuc) [inline test/common/stats/thread_local_store_test.cc]: @mrice32 I was able to completely swap in this TestAllocator for a real, functional BlockRawStatDataAllocator, which ends up running the below suite of namespace resolution tests.

**Comment by ambuc** [comment]
Closed in favor of a PR upstream.
